presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java

```diff
@@ -163,4 +181,89 @@ public static int checkedAdd(int x, int y)
             throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Concatenated string is too large");
         }
     }
+
+    private static class ConcatSignatureBinder
+            extends SignatureBinder
+    {
+        public ConcatSignatureBinder(TypeManager typeManager, Signature signature, boolean allowCoercion)
+        {
+            super(typeManager, signature, allowCoercion);
+        }
+
+        @Override
+        public Optional<Signature> bind(List<? extends Type> actualArgumentTypes)
+        {
+            ImmutableList.Builder<TypeSignature> boundTypesSignatures = ImmutableList.builder();
+
+            long lengthAccumulator = 0;
+            for (Type type : actualArgumentTypes) {
+                if (!isTypeValid(type)) {
+                    return Optional.empty();
+                }
+
+                int paramLength = varcharLength(type);
+                lengthAccumulator = Ints.saturatedCast(lengthAccumulator + paramLength);
+                boundTypesSignatures.add(VarcharType.createVarcharType(paramLength).getTypeSignature());
+            }
+
+            return Optional.of(
+                    new Signature(
+                            declaredSignature.getName(),
+                            declaredSignature.getKind(),
+                            VarcharType.createVarcharType(Ints.saturatedCast(lengthAccumulator)).getTypeSignature(),
+                            boundTypesSignatures.build()));
+        }
+
+        @Override
+        public Optional<BoundVariables> bindVariables(List<? extends Type> actualArgumentTypes, Type actualReturnType)
+        {
+            Optional<Signature> signature = bind(actualArgumentTypes);
```

Comment:
Why return any map with bound variables while input signature have no variables at all.
You just put values for `x<num>` and `r` - and those will never be used later on anyway.
Right?